File: rust/plugins/sensors/ios/Sources/SensorsPlugin.swift
import CoreMotion
import Foundation
import SwiftRs
import Tauri
import WebKit

struct ReadArgs: Decodable {
    let sensor: String
    let id: String?
}

struct WatchArgs: Decodable {
    let sensor: String
    let intervalMs: Double?
    let id: String?
}

/**
 * Raw sensor access via CoreMotion. Units follow the Expo convention:
 * accelerometer in g (CoreMotion native), gyroscope in rad/s, magnetometer
 * in μT, barometer in hPa (CMAltimeter reports kPa, converted). The ambient
 * light sensor has no public API on iOS — `light` reports unavailable.
 * No permissions: raw readings are unrestricted (NSMotionUsageDescription
 * is only needed by CMPedometer/DeviceMotion activity, which are not used).
 */
class SensorsPlugin: Plugin {
    private let motion = CMMotionManager()
    private let altimeter = CMAltimeter()
    private let queue = OperationQueue()
    // Sensors with an active watch; one-shots piggyback on the same underlying
    // CoreMotion stream and stop it unless a watch owns it.
    private var watching = Set<String>()

    @objc public func isAvailable(_ invoke: Invoke) {
        guard let args = try? invoke.parseArgs(ReadArgs.self) else {
            invoke.reject("invalid args")
            return
        }
        let ok: Bool
        switch args.sensor {
        case "accelerometer": ok = motion.isAccelerometerAvailable
        case "gyroscope": ok = motion.isGyroAvailable
        case "magnetometer": ok = motion.isMagnetometerAvailable
        case "barometer": ok = CMAltimeter.isRelativeAltitudeAvailable()
        default: ok = false
        }
        invoke.resolve(["sensor": args.sensor, "id": args.id as Any, "available": ok])
    }

    @objc public func readSensor(_ invoke: Invoke) {
        guard let args = try? invoke.parseArgs(ReadArgs.self) else {
            invoke.reject("invalid args")
            return
        }
        let started = start(args.sensor, id: args.id, intervalMs: 100, once: true) { [weak self] payload in
            if !(self?.watching.contains(args.sensor) ?? false) {
                self?.stop(args.sensor)
            }
            invoke.resolve(payload ?? Self.unavailable(args.sensor, args.id))
        }
        if !started {
            // Unsupported/missing sensor: start() never schedules the 2s
            // timeout, so answer here or the JS promise hangs forever.
            invoke.resolve(Self.unavailable(args.sensor, args.id))
        }
    }

    @objc public func watchSensor(_ invoke: Invoke) {
        guard let args = try? invoke.parseArgs(WatchArgs.self) else {
            invoke.reject("invalid args")
            return
        }
        let interval = max(args.intervalMs ?? 500, 100) // floor: PHP round-trips
        watching.insert(args.sensor)
        let started = start(args.sensor, id: args.id, intervalMs: interval, once: false) { [weak self] payload in
            if let payload = payload {
                self?.emit(payload)
            }
        }
        if started {
            invoke.resolve(["watching": true])
        } else {
            watching.remove(args.sensor)
            invoke.resolve(Self.unavailable(args.sensor, args.id))
        }
    }

    @objc public func stopSensor(_ invoke: Invoke) {
        guard let args = try? invoke.parseArgs(ReadArgs.self) else {
            invoke.reject("invalid args")
            return
        }
        watching.remove(args.sensor)
        stop(args.sensor)
        invoke.resolve()
    }

    @discardableResult
    private func start(
        _ sensor: String,
        id: String?,
        intervalMs: Double,
        once: Bool,
        handler: @escaping ([String: Any]?) -> Void
    ) -> Bool {
        let interval = intervalMs / 1000.0
        var fired = false
        let deliver: ([String: Any]?) -> Void = { payload in
            if once {
                if fired { return }
                fired = true
            }
            handler(payload)
        }

        switch sensor {
        case "accelerometer":
            guard motion.isAccelerometerAvailable else { return false }
            motion.accelerometerUpdateInterval = interval
            motion.startAccelerometerUpdates(to: queue) { data, _ in
                guard let a = data?.acceleration else { return }
                deliver(Self.xyz(sensor, id, a.x, a.y, a.z))
            }
        case "gyroscope":
            guard motion.isGyroAvailable else { return false }
            motion.gyroUpdateInterval = interval
            motion.startGyroUpdates(to: queue) { data, _ in
                guard let r = data?.rotationRate else { return }
                deliver(Self.xyz(sensor, id, r.x, r.y, r.z))
            }
        case "magnetometer":
            guard motion.isMagnetometerAvailable else { return false }
            motion.magnetometerUpdateInterval = interval
            motion.startMagnetometerUpdates(to: queue) { data, _ in
                guard let f = data?.magneticField else { return }
                deliver(Self.xyz(sensor, id, f.x, f.y, f.z))
            }
        case "barometer":
            guard CMAltimeter.isRelativeAltitudeAvailable() else { return false }
            altimeter.startRelativeAltitudeUpdates(to: queue) { data, _ in
                guard let kpa = data?.pressure.doubleValue else { return }
                deliver([
                    "sensor": sensor,
                    "id": id as Any,
                    "available": true,
                    "value": kpa * 10.0, // kPa → hPa
                    "timestamp": Int(Date().timeIntervalSince1970),
                ])
            }
        default:
            return false // "light" has no public API on iOS
        }

        if once {
            // Never leak the invoke if the sensor stays silent.
            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                deliver(nil)
            }
        }
        return true
    }

    private func stop(_ sensor: String) {
        switch sensor {
        case "accelerometer": motion.stopAccelerometerUpdates()
        case "gyroscope": motion.stopGyroUpdates()
        case "magnetometer": motion.stopMagnetometerUpdates()
        case "barometer": altimeter.stopRelativeAltitudeUpdates()
        default: break
        }
    }

    private func emit(_ payload: [String: Any]) {
        var data = JSObject()
        for (k, v) in payload {
            if let b = v as? Bool { data[k] = b }
            else if let d = v as? Double { data[k] = d }
            else if let i = v as? Int { data[k] = i }
            else if let s = v as? String { data[k] = s }
        }
        trigger("sensor-changed", data: data)
    }

    private static func xyz(_ sensor: String, _ id: String?, _ x: Double, _ y: Double, _ z: Double) -> [String: Any] {
        return [
            "sensor": sensor,
            "id": id as Any,
            "available": true,
            "x": x,
            "y": y,
            "z": z,
            "timestamp": Int(Date().timeIntervalSince1970),
        ]
    }

    private static func unavailable(_ sensor: String, _ id: String? = nil) -> [String: Any] {
        return ["sensor": sensor, "id": id as Any, "available": false]
    }
}

@_cdecl("init_plugin_nativeblade_sensors")
func initPlugin() -> Plugin {
    return SensorsPlugin()
}

File: rust/plugins/sensors/android/src/main/java/app/nativeblade/sensors/SensorsPlugin.kt
package app.nativeblade.sensors

import android.app.Activity
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.SystemClock
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin

@InvokeArg
class ReadArgs {
    lateinit var sensor: String
    var id: String? = null
}

@InvokeArg
class WatchArgs {
    lateinit var sensor: String
    var intervalMs: Long = 500
    var id: String? = null
}

/**
 * Raw sensor access. Units follow the Expo convention devs already know:
 * accelerometer in g, gyroscope in rad/s, magnetometer in μT, barometer in
 * hPa, light in lux. Watches are throttled JS-side friendly: the OS listener
 * may fire faster, but events are forwarded at most once per interval.
 */
@TauriPlugin
class SensorsPlugin(private val activity: Activity) : Plugin(activity) {

    private val manager: SensorManager
        get() = activity.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    // One active watch listener per sensor type.
    private val watchers = HashMap<String, SensorEventListener>()

    @Command
    fun isAvailable(invoke: Invoke) {
        val args = invoke.parseArgs(ReadArgs::class.java)
        val obj = JSObject()
        obj.put("sensor", args.sensor)
        obj.put("id", args.id)
        obj.put("available", resolve(args.sensor) != null)
        invoke.resolve(obj)
    }

    @Command
    fun readSensor(invoke: Invoke) {
        val args = invoke.parseArgs(ReadArgs::class.java)
        val sensor = resolve(args.sensor) ?: run {
            invoke.resolve(unavailable(args.sensor, args.id))
            return
        }

        // One-shot: register, take the first sample, unregister.
        var done = false
        val listener = object : SensorEventListener {
            override fun onSensorChanged(event: SensorEvent) {
                if (done) return
                done = true
                manager.unregisterListener(this)
                invoke.resolve(payload(args.sensor, args.id, event))
            }

            override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
        }
        manager.registerListener(listener, sensor, SensorManager.SENSOR_DELAY_UI)

        // A sensor that never fires (rare, but light sensors only report on
        // change) must not leak the invoke: give it two seconds.
        android.os.Handler(activity.mainLooper).postDelayed({
            if (!done) {
                done = true
                manager.unregisterListener(listener)
                invoke.resolve(unavailable(args.sensor, args.id))
            }
        }, 2000)
    }

    @Command
    fun watchSensor(invoke: Invoke) {
        val args = invoke.parseArgs(WatchArgs::class.java)
        val sensor = resolve(args.sensor) ?: run {
            invoke.resolve(unavailable(args.sensor, args.id))
            return
        }
        val interval = args.intervalMs.coerceAtLeast(100) // floor: PHP round-trips

        stop(args.sensor)
        var lastEmit = 0L
        val listener = object : SensorEventListener {
            override fun onSensorChanged(event: SensorEvent) {
                val now = SystemClock.elapsedRealtime()
                if (now - lastEmit < interval) return
                lastEmit = now
                trigger("sensor-changed", payload(args.sensor, args.id, event))
            }

            override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
        }
        watchers[args.sensor] = listener
        manager.registerListener(listener, sensor, (interval * 1000).toInt())

        val ok = JSObject()
        ok.put("watching", true)
        invoke.resolve(ok)
    }

    @Command
    fun stopSensor(invoke: Invoke) {
        val args = invoke.parseArgs(ReadArgs::class.java)
        stop(args.sensor)
        invoke.resolve()
    }

    private fun stop(type: String) {
        watchers.remove(type)?.let { manager.unregisterListener(it) }
    }

    private fun resolve(type: String): Sensor? = when (type) {
        "accelerometer" -> manager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        "gyroscope" -> manager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
        "magnetometer" -> manager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
        "barometer" -> manager.getDefaultSensor(Sensor.TYPE_PRESSURE)
        "light" -> manager.getDefaultSensor(Sensor.TYPE_LIGHT)
        else -> null
    }

    private fun payload(type: String, id: String?, event: SensorEvent): JSObject {
        val obj = JSObject()
        obj.put("sensor", type)
        obj.put("id", id)
        obj.put("available", true)
        obj.put("timestamp", System.currentTimeMillis() / 1000)
        when (type) {
            // Android reports m/s²; Expo-style g for portability.
            "accelerometer" -> {
                obj.put("x", event.values[0] / SensorManager.GRAVITY_EARTH)
                obj.put("y", event.values[1] / SensorManager.GRAVITY_EARTH)
                obj.put("z", event.values[2] / SensorManager.GRAVITY_EARTH)
            }
            "gyroscope", "magnetometer" -> {
                obj.put("x", event.values[0])
                obj.put("y", event.values[1])
                obj.put("z", event.values[2])
            }
            "barometer" -> obj.put("value", event.values[0])
            "light" -> obj.put("value", event.values[0])
        }
        return obj
    }

    private fun unavailable(type: String, id: String? = null): JSObject {
        val obj = JSObject()
        obj.put("sensor", type)
        obj.put("id", id)
        obj.put("available", false)
        return obj
    }
}
